File: src/com/castlefrog/agl/agents/RandomAgent.java
package com.castlefrog.agl.agents;

import java.util.List;

import com.castlefrog.agl.Action;
import com.castlefrog.agl.Agent;
import com.castlefrog.agl.Simulator;
import com.castlefrog.agl.State;

/**
 * This agent selects a random action from the list
 * of possible actions from a given state.
 */
public final class RandomAgent implements Agent {
    public RandomAgent() {
    }

    public <S extends State<S>, A extends Action> A selectAction(int agentId, S state, Simulator<S, A> simulator) {
        simulator.setState(state);
        List<A> actions = simulator.getLegalActions(agentId);
        return actions.get((int) (Math.random() * actions.size()));
    }

    public String getName() {
        return "random";
    }

    @Override
    public String toString() {
        return getName() + " agent";
    }
}

File: src/main/kotlin/com/castlefrog/agl/agents/RandomAgent.kt
package com.castlefrog.agl.agents

import arrow.core.Either
import com.castlefrog.agl.*
import com.castlefrog.agl.util.getPlayerActions
import kotlin.random.Random

/**
 * This agent selects an action at random (normal distribution)
 * from the list of possible actions from a given state.
 */
class RandomAgent(private val random: Random = Random) : Agent {

    override fun <S : State<S>, A : Action<A>> selectAction(
        playerId: Int,
        state: S,
        simulator: Simulator<S, A>
    ): Either<ResultError, A> {
        val result = simulator.calculateLegalActions(state).getPlayerActions(playerId)
        return result.fold({
            Either.Left(it)
        }, {
            Either.Right(it.random(random))
        })
    }

    override fun toString(): String = RandomAgent::class.java.simpleName
}
